<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [Mono-bugs] [Bug 82034][Nor] New - goto does not call the finally	block to close the iterator when leaving a foreach loop
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:mono-bugs%40lists.ximian.com?Subject=%5BMono-bugs%5D%20%5BBug%2082034%5D%5BNor%5D%20New%20-%20goto%20does%20not%20call%20the%20finally%0A%09block%20to%20close%20the%20iterator%20when%20leaving%20a%20foreach%20loop&In-Reply-To=bug-82034%40chernobyl.ximian.com">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="058636.html">
   <LINK REL="Next"  HREF="058638.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[Mono-bugs] [Bug 82034][Nor] New - goto does not call the finally	block to close the iterator when leaving a foreach loop</H1>
    <B>bugzilla-daemon at bugzilla.ximian.com</B> 
    <A HREF="mailto:mono-bugs%40lists.ximian.com?Subject=%5BMono-bugs%5D%20%5BBug%2082034%5D%5BNor%5D%20New%20-%20goto%20does%20not%20call%20the%20finally%0A%09block%20to%20close%20the%20iterator%20when%20leaving%20a%20foreach%20loop&In-Reply-To=bug-82034%40chernobyl.ximian.com"
       TITLE="[Mono-bugs] [Bug 82034][Nor] New - goto does not call the finally	block to close the iterator when leaving a foreach loop">bugzilla-daemon at bugzilla.ximian.com
       </A><BR>
    <I>Sat Jul  7 13:21:37 EDT 2007</I>
    <P><UL>
        <LI>Previous message: <A HREF="058636.html">[Mono-bugs] [Bug 82012][Nor] Changed - mod_mono not serving requests
</A></li>
        <LI>Next message: <A HREF="058638.html">[Mono-bugs] [Bug 81982][Nor] Changed - Failing to call EndInvoke on	a delegate causes apps to hang on exit
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#58637">[ date ]</a>
              <a href="thread.html#58637">[ thread ]</a>
              <a href="subject.html#58637">[ subject ]</a>
              <a href="author.html#58637">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Please do not reply to this email- if you want to comment on the bug, go to the
URL shown below and enter your comments there.

Changed by <A HREF="http://lists.ximian.com/mailman/listinfo/mono-bugs">jeff at thefirst.org.</A>

<A HREF="http://bugzilla.ximian.com/show_bug.cgi?id=82034">http://bugzilla.ximian.com/show_bug.cgi?id=82034</A>

--- shadow/82034	2007-07-07 13:21:37.000000000 -0400
+++ shadow/82034.tmp.4264	2007-07-07 13:21:37.000000000 -0400
@@ -0,0 +1,86 @@
+Bug#: 82034
+Product: Mono: Compilers
+Version: 1.2
+OS: 
+OS Details: XP
+Status: NEW   
+Resolution: 
+Severity: 
+Priority: Normal
+Component: C#
+AssignedTo: <A HREF="http://lists.ximian.com/mailman/listinfo/mono-bugs">rharinath at novell.com</A>                            
+ReportedBy: <A HREF="http://lists.ximian.com/mailman/listinfo/mono-bugs">jeff at thefirst.org</A>               
+QAContact: <A HREF="http://lists.ximian.com/mailman/listinfo/mono-bugs">mono-bugs at ximian.com</A>
+TargetMilestone: ---
+URL: 
+Cc: 
+Summary: goto does not call the finally block to close the iterator when leaving a foreach loop
+
+Please fill in this template when reporting a bug, unless you know what you
+are doing.
+Description of Problem:
+Usually, in an iterator, we can put yield inside a try/finally block and
+when control leaves the foreach loop for the iterator (because of &quot;break&quot;
+or an exception) it will call the iterator's finally block. But if we leave
+the foreach loop using goto, it doesn't do this. But we use &quot;goto&quot; to
+transfer control out of one or more nested loops, and it should behave the
+same as &quot;break&quot;.
+
+Steps to reproduce the problem:
+1. Compile and run this code:
+using System.Collections.Generic;
+
+class TestGoto
+{
+    static int x = 0;
+
+    static void Main(string[] args)
+    {
+        foreach (bool b in test())
+        {
+        }
+    }
+
+    static IEnumerable&lt;bool&gt; setX()
+    {
+        x = 1;
+        try
+        {
+            yield return true;
+        }
+        finally
+        {
+            x = 0;
+        }
+    }
+
+    static IEnumerable&lt;bool&gt; test()
+    {
+        foreach (bool b in setX())
+        {
+            yield return true;
+            // Change &quot;goto label&quot; to &quot;break&quot; to show the correct result.
+            goto label;
+        }
+    label:
+        if (x == 0)
+            System.Console.WriteLine(&quot;Success: finally was called&quot;);
+        else
+            System.Console.WriteLine(&quot;Error: finally was not called&quot;);
+    }
+}
+
+
+Actual Results:
+prints &quot;Error: finally was not called&quot; because the setX() iterator is not
+closed when leaving the foreach loop.
+
+Expected Results:
+If you change &quot;goto label&quot; to &quot;break&quot;, then the iterator is closed properly
+and it prints &quot;Success: finally was called&quot;. But a &quot;goto&quot; should act like
+&quot;break&quot; and also close the iterator.
+
+How often does this happen? 
+Every time
+
+Additional Information:
</PRE>


<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="058636.html">[Mono-bugs] [Bug 82012][Nor] Changed - mod_mono not serving requests
</A></li>
	<LI>Next message: <A HREF="058638.html">[Mono-bugs] [Bug 81982][Nor] Changed - Failing to call EndInvoke on	a delegate causes apps to hang on exit
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#58637">[ date ]</a>
              <a href="thread.html#58637">[ thread ]</a>
              <a href="subject.html#58637">[ subject ]</a>
              <a href="author.html#58637">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="http://lists.ximian.com/mailman/listinfo/mono-bugs">More information about the mono-bugs
mailing list</a><br>
</body></html>
